packages/jest-runtime/src/index.ts

```diff
@@ -758,17 +819,42 @@ export default class Runtime {
     }
 
     if (module.status === 'unlinked') {
+      const preserveLoadOrder =
+        this._config.preserveLoadOrder || this._globalConfig?.preserveLoadOrder;
+      // ensure that every import fully evaluates before any siblings are allowed to import.
+      let linkPromiseChain = (Promise<VMModule | void>).resolve();
       // since we might attempt to link the same module in parallel, stick the promise in a weak map so every call to
       // this method can await it
       this._esmModuleLinkingMap.set(
         module,
-        module.link((specifier: string, referencingModule: VMModule) =>
-          this.resolveModule(
+        module.link((specifier: string, referencingModule: VMModule) => {
+          if (preserveLoadOrder) {
+            linkPromiseChain = linkPromiseChain.then(async () => {
+              const mod = await this.resolveModule<VMModule>(
+                specifier,
+                referencingModule.identifier,
+                referencingModule.context,
+              );
+              const hasCyclicDependency = await this.checkForCyclicDependencies(
+                mod,
+                parentChain,
+              );
+              if (mod.status === 'unlinked' && !hasCyclicDependency) {
+                await this.linkAndEvaluateModule(
+                  mod,
+                  new Set<VMModule>([...parentChain, mod]),
+                );
+              }
+              return mod;
```

Comment:
can you extract this into a separate method?

Reply:
certainly!